Forbid the removal of the write block if the index is read-only

tlrx opened the pull request:
Ensure that a `write` block cannot be removed on a read-only compatible index in version N-2, while allowing to change a `read_only` block into a `write` block if needed as well as closing/reopening such indices.

Requires #120647 to be merged on `8.x`.

Relates ES-10320

tlrx committed:
Forbid the removal of the write block if the index is read-only

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:
```diff
@@ -443,6 +453,17 @@ private static boolean maybeUpdateClusterBlock(
                         changed = true;
                         if (block.contains(ClusterBlockLevel.WRITE)) {
```
I think this is too harsh. In case the index has both the read-only and the write block, it should be allowed to remove either.

In particular, if the user applied read-only on 8.x, upgraded and then want to update some settings, they should be allowed to add the write block and remove the read-only block.

I recognize this is also a flaw in the original code in that it removes the verified setting.

I could be ok to let this go in if we follow-up with the additional detail. But would prefer we do it in one go if it is not too hard.

tlrx replied:
> I think this is too harsh. In case the index has both the read-only and the write block, it should be allowed to remove either.

I agree, it makes sense.

>In particular, if the user applied read-only on 8.x, upgraded and then want to update some settings, they should be allowed to add the write block and remove the read-only block.

I initially thought that updating both settings at once would not work because of the code allowing setting updates when there is a `METADATA_WRITE` block in place:
https://github.com/elastic/elasticsearch/blob/45ae0718cc6f52c78c9dc96380d623be9917ba07/server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java#L87-L91

but it seems that `request.settings().size() == 1` does not always apply since the great setting refactoring [here](https://github.com/elastic/elasticsearch/commit/eb7012101acd4902d65e5a438fac876a3153868a#diff-c68d5ecca4e08cd390830bbfa43e307ccc436e1a8a821d2c57c69ce357ad5fc9R65).

It looks like this refactoring introduced a change in the parenthesis impacting the conditions to bypass the check. Which will serve us today.

henningandersen reviewed (commented):
Looks good though I hope we can support removing one of two write blocks too? See comment.

tlrx committed:
Merge branch 'main' into 2025/01/22/forbid-write-block-removal

tlrx committed:
Merge branch 'main' into 2025/01/22/forbid-write-block-removal

tlrx committed:
rework block & settings handling

tlrx committed:
Merge branch 'main' into 2025/01/22/forbid-write-block-removal

tlrx commented:
Ok, so it took me a couple of days to get this working with the subtleties of blocks updates. Thanks for your patience on this @henningandersen. 

I updated this PR (and #120647) to forbid the removal of the `ClusterBlockLevel.WRITE` on a read-only compatible index if at least one node of the cluster cannot write to it (see `verifyReadOnlyIndices`). 

Since both `index.blocks.read_only` to `index.blocks.write` contains this `ClusterBlockLevel.WRITE` we can allow updating the index blocks setting to move from one type of block to the other. In the tests I included the use case you mentioned: updating from `index.blocks.read_only` to `index.blocks.write`.

I also changed the `MetadataUpdateSettingsService#maybeUpdateClusterBlock` so that the `VERIFIED_READ_ONLY_SETTING` remains as long as a `ClusterBlockLevel.WRITE` remains (required for blocks settings updates). It also allows to keep the VERIFIED_READ_ONLY_SETTING value when closing a verified read-only index, because closed indices have the INDEX_CLOSED_BLOCK that also contains the `ClusterBlockLevel.WRITE` block. When reopening, the INDEX_CLOSED_BLOCK is removed, removing the VERIFIED_READ_ONLY_SETTING if no other write block is in place and it will fail the reopening since the check in `verifyReadOnlyIndices` does not allow a read-only compatible index to have no `ClusterBlockLevel.WRITE` block (applying a write block before reopening will work).

I largely updated the integration tests to to have a bit more randomization around blocks and closing/opening during upgrades.

Note that there is still a bit missing for fully supporting closed indices (allowing a verified-before-close index to be upgraded without being marked as read-only first): I'll do this in follow ups.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:
```diff
@@ -417,6 +441,24 @@ public static void updateIndexSettings(
+    private static void verifyReadOnlyIndices(@Nullable Set<Index> readOnlyIndices, ClusterBlocks blocks) {
+        if (readOnlyIndices != null) {
+            for (Index readOnlyIndex : readOnlyIndices) {
+                if (blocks.indexBlocked(ClusterBlockLevel.WRITE, readOnlyIndex.getName()) == false) {
```
Should this use the new `hasIndexBlockLevel` method to avoid interaction from global blocks?

tlrx replied:
Right, I pushed de6a0bdb02283850083ed6d633577643dd5f33fb.

henningandersen commented on qa/lucene-index-compatibility/src/javaRestTest/java/org/elasticsearch/lucene/RollingUpgradeLuceneIndexCompatibilityTestCase.java:
```diff
@@ -194,6 +249,15 @@ public void testRestoreIndex() throws Exception {
                 closeIndex(restoredIndex);
                 ensureGreen(restoredIndex);
 
+                logger.debug("--> write API block can be removed on a closed index: INDEX_CLOSED_BLOCK already blocks writes");
```
Can we also check that if an index has both read-only and write block then we can remove one of them for an N-2 index?

tlrx replied:
I reworked the testIndexUpgrade test to randomly add one or two of the blocks, and then remove one of them.

See 59e7acddb8addaeda32172a3ae90ece223f6cde4

henningandersen reviewed (approved):
LGTM.

tlrx committed:
Merge branch 'main' into 2025/01/22/forbid-write-block-removal

tlrx committed:
feedback

tlrx committed:
feedback

tlrx committed:
Merge branch 'main' into 2025/01/22/forbid-write-block-removal

tlrx commented:
Thanks Henning
